crates/bevy_asset/src/asset_changed.rs

```diff
@@ -0,0 +1,475 @@
+/// [`Assets<Mesh>::get_mut`] method, `AssetChanged<Mesh>` will iterate over all
+/// entities with the `Handle<Mesh>` for that mesh. Meanwhile, `Changed<Handle<Mesh>>`
+/// will iterate over no entities.
+///
+/// You probably need to check for both asset and handle changes. If this is the case,
+/// use [`AssetOrHandleChanged`] instead.
+///
+/// # Quirks
+///
+/// - Asset changes are registered in the [`AssetEvents`] schedule.
+/// - Removed assets are not detected.
+/// - Asset update tracking only starts after the first system with an
+///   `AssetChanged` query parameter ran for the first time.
+///
+/// This means that assets added before the system ran won't be detected
+/// (for example in a `Startup` schedule).
+///
+/// This is also true of systems gated by a `run_if` condition.
+///
+/// The list of changed assets only gets updated in the
+/// [`AssetEvents`] schedule, just after `PostUpdate`. Therefore, `AssetChanged`
+/// will only pick up asset changes in schedules following `AssetEvents` or the
+/// next frame. Consider adding the system in the `Last` schedule if you need
+/// to react without frame delay to asset changes.
+///
+/// # Performance
+///
+/// When at least one `A` is updated, this will
+/// read a hashmap once per entity with a `Handle<A>` component. The
+/// runtime of the query is proportional to how many entities with a `Handle<A>`
+/// it matches.
+///
+/// If no `A` asset updated since the last time the system ran, then no lookups occur.
+///
+/// [`AssetEvents`]: crate::AssetEvents
+/// [`Assets<Mesh>::get_mut`]: crate::Assets::get_mut
+pub struct AssetChanged<A: Asset>(PhantomData<A>);
+
+/// Fetch for [`AssetChanged`].
+#[doc(hidden)]
+pub struct AssetChangedFetch<'w, A: Asset> {
+    inner: Option<ReadFetch<'w, Handle<A>>>,
+    check: AssetChangeCheck<'w, A>,
+}
+impl<'w, A: Asset> Clone for AssetChangedFetch<'w, A> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner,
+            check: self.check,
+        }
+    }
+}
+
+/// State for [`AssetChanged`].
+#[doc(hidden)]
+pub struct AssetChangedState<A: Asset> {
+    handle_id: ComponentId,
+    resource_id: ComponentId,
+    archetype_id: ArchetypeComponentId,
+    _asset: PhantomData<fn(A)>,
+}
+
+// SAFETY: `ROQueryFetch<Self>` is the same as `QueryFetch<Self>`
+unsafe impl<A: Asset> WorldQuery for AssetChanged<A> {
+    type ReadOnly = Self;
+    type Fetch<'w> = AssetChangedFetch<'w, A>;
+    type State = AssetChangedState<A>;
+
+    type Item<'w> = bool;
+
+    fn shrink<'wlong: 'wshort, 'wshort>(item: QueryItem<'wlong, Self>) -> QueryItem<'wshort, Self> {
+        item
+    }
+
+    fn init_state(world: &mut World) -> AssetChangedState<A> {
+        let resource_id = world.init_resource::<AssetChanges<A>>();
+        let archetype_id = world.storages().resources.get(resource_id).unwrap().id();
+        AssetChangedState {
+            handle_id: world.init_component::<Handle<A>>(),
```

Comment:
The documentation for `World::init_component()` doesn't specify what happens if the component is already initialized (unlike `init_resource()` which is a no-op in that case). Is it safe to unconditionally init here?